<img src="https://images.creativefabrica.com/products/thumbnails/2023/12/16/C7T2JTITP/2Zcuu0O3O4fe8oV5YBepq5FPXdj.png"
     alt="Porquinhos Felizes" style="display:block;margin:0 auto;width:30%;">


# Porquinhos Felizes

**Porquinhos Felizes** é uma pequena fazenda de porcos. Entre os desafios que eles enfrentam para criar animais saudáveis está o de compor uma **dieta nutritiva o mais barata possível.**

Uma boa dieta suína deve **conter quantidade suficiente de energia e proteína.**

A fazenda depende de uma mistura de **arroz e milho** para fornecer esses nutrientes.

As tabelas abaixo contêm o **custo** de cada ingrediente, a ingestão diária **mínima e máxima** recomendada de cada nutriente e a **quantidade** de cada nutriente em cada ingrediente.

*   Alimentos

<table>
<thead>
  <tr align="center">
    <th>Alimento</th>
    <th>Custo (R$/Kg)</th>
  </tr>
</thead>
<tbody>
  <tr align="center">
    <td>arroz</td>
    <td>1,50</td>
  </tr>
  <tr align="center">
    <td>milho</td>
    <td>2,23</td>
  </tr>
</tbody>
</table>

*   Nutrientes

<table>
<thead>
  <tr align="center">
    <th>Nutriente</th>
    <th>Ingestão Mínima (Kg)</th>
    <th>Ingestão Máxima (Kg)</th>
  </tr>
</thead>
<tbody>
  <tr align="center">
    <td>carboidratos</td>
    <td>6,0</td>
    <td>7,5</td>
  </tr>
  <tr align="center">
    <td>proteínas</td>
    <td>0,9</td>
    <td>1,6</td>
  </tr>
</tbody>
</table>

*   Nutrientes por Alimento

<table>
<thead>
  <tr align="center">
    <th>Alimento</th>
    <th>Nutriente</th>
    <th>Quantidade (Kg)</th>
  </tr>
</thead>
<tbody>
  <tr align="center">
    <td>arroz</td>
    <td>carboidratos</td>
    <td>0,77</td>
  </tr>
  <tr align="center">
    <td>arroz</td>
    <td>proteínas</td>
    <td>0,08</td>
  </tr>
  <tr align="center">
    <td>milho</td>
    <td>carboidratos</td>
    <td>0,66</td>
  </tr>
  <tr align="center">
    <td>milho</td>
    <td>proteínas</td>
    <td>0,14</td>
  </tr>
</tbody>
</table>

# Formulação

## Variáveis de decisão

x1 = quantidade de arroz a ser consumida em kg, 
x2 = quantidade de milho a ser consumida em kg.

- O que deve ser decidido é a quantidade de consumo de cada alimento.
- O objetivo é minimizar o custo da dieta.

Podemos calcular o custo da dieta com a seguinte função:

Custo = 1,50 * x1 + 2,23 * x2

--

arroz_custo = 1,50
milho_custo = 2,23

arroz_carboidratos = 0,77
arroz_proteinas = 0,08
milho_carboidratos = 0,66
milho_proteinas = 0,14









## Restrições

A quantidade de nutrientes na dieta deve estar dentro dos limites recomendados.

As restrições impostas são:

- A quantidade de **carboidratos** deve ser entre 6.0 e 7.5 kg.
- A quantidade de **proteínas** deve ser entre 0.9 e 1.6 kg.
- A quantidade de arroz e milho deve ser maior ou igual a 0.

Representando matematicamente:

6.0 <= (0.77 * x[1] + 0.66 * x[2]) <= 7.5  
0.9 <= (0.08 * x[1] + 0.14 * x[2]) <= 1.6
x[1], x[2] >= 0


## Função objetivo

Objetivo é minimizar o custo da dieta.

Custo = 1,50 * x1 + 2,23 * x2

## Formulação Final

In [1]:
!pip install pulp

In [28]:
import pulp

# definir → restringir → resolver → ler solução

In [37]:
# Modelo
modelo = pulp.LpProblem(
    name="problema_dos_porquinhos_felizes",
    sense=pulp.LpMinimize
)

# Queremos minimizar o custo da dieta para que os porquinhos fiquem felizes
# e saudáveis, pela escolha ótima de nutrientes

In [40]:
# Definir variaveis

# x1 = arroz
# x2 = milho

x = pulp.LpVariable.dicts(
    indices=["arroz","milho"],
    cat=pulp.LpContinuous,
    lowBound=0,
    name='x'
)

In [43]:
# Definir função objetiva

modelo.setObjective(1.50 * x["arroz"] + 2.23 * x["milho"])

In [46]:
# Definir as restrições

modelo.addConstraint(0.77 * x["arroz"] + 0.66 * x["milho"] <= 7.5, name='dieta_carboidratos_maximo')
modelo.addConstraint(0.77 * x["arroz"] + 0.66 * x["milho"] >= 6.0, name='dieta_carboidratos_minimo')
modelo.addConstraint(0.08 * x["arroz"] + 0.14 * x["milho"] >= 0.9, name='dieta_proteinas_minimo')
modelo.addConstraint(0.08 * x["arroz"] + 0.14 * x["milho"] <= 1.6, name='dieta_proteinas_maximo')


In [49]:
# Resultado
modelo.solve()

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /root/venv/lib/python3.12/site-packages/pulp/apis/../solverdir/cbc/linux/i64/cbc /tmp/75b5e9e148934cc194f1d2cd17dab060-pulp.mps -timeMode elapsed -branch -printingOptions all -solution /tmp/75b5e9e148934cc194f1d2cd17dab060-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 9 COLUMNS
At line 20 RHS
At line 25 BOUNDS
At line 26 ENDATA
Problem MODEL has 4 rows, 2 columns and 8 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 4 (0) rows, 2 (0) columns and 8 (0) elements
0  Obj 0 Primal inf 14.220779 (2)
2  Obj 15.345273
Optimal - objective value 15.345273
Optimal objective 15.34527273 - 2 iterations time 0.002
Option for printingOptions changed from normal to all
Total time (CPU seconds):       0.00   (Wallclock seconds):       0.00



1

In [73]:
# Relatório

for variables in modelo.variables():
    print(f"{variables.name}: {variables.varValue}")

print(f"Custo total: {pulp.value(modelo.objective)}")
print(f"Status: {pulp.LpStatus[modelo.status]}")

x_arroz: 4.4727273
x_milho: 3.8727273
Custo total: 15.345272829
Status: Optimal


### Conclusão

<p>A dieta de menor custo para deixar os porquinhos felizes é de 4.47 kg de arroz e 3.87 kg de milho por dia, atendendo aos mínimos de carboidratos (6.0 kg) e proteínas (0.9 kg) e respeitando os máximos, com custo total de R$15.345 por dia.</p>

<p> O arroz é mais barato (R$1,50 por kg vs R$2.23 por kg), mas o milho tem mais proteína por kg (0,14 vs 0,08). A solução ótima é uma combinação dos dois alimentos para minimizar o custo e atender às necessidades nutricionais. </p>

<p>Milho sozinho seria viável mas muito mais caro (~R$20 e poucos por dia vs R$15.35)</p>